File: lecture_0-2/assignements/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int letter_count(string s);
int word_count(string s);
int sentence_count(string s);
void print_grade(int index);

int main(void)
{
    // Get the text.
    string text = get_string("Text: ");

    // Calculate the index.
    int letters = letter_count(text);
    int words = word_count(text);
    int sentences = sentence_count(text);

    // Calculate the index of readability knowing that L is the average number of letters per 100
    // words and S is the average number of sentences per 100 words.
    float L = letters * 100.0 / words;
    float S = sentences * 100.0 / words;
    float index = 0.0588 * L - 0.295 * S - 15.8;

    // Print the grade.
    print_grade(round(index));
    return 0;
}

int letter_count(string s)
{
    int i = 0;
    int length = 0;

    // Count letters until the input is over.
    while (s[i] != '\0')
    {
        if (isalnum(s[i]))
        {
            length++;
        }
        i++;
    }
    return length;
}

int word_count(string s)
{
    int i = 0;
    int word = 1;

    // Count words until the input is over.
    while (s[i] != '\0')
    {
        if (isspace(s[i]))
        {
            word++;
        }
        i++;
    }
    return word;
}

int sentence_count(string s)
{
    int i = 0;
    int sentence = 0;

    // Count sentences until the input is over.
    while (s[i] != '\0')
    {
        if (s[i] == '.' || s[i] == '?' || s[i] == '!') // || s[i] == ';')
        {
            sentence++;
        }
        i++;
    }
    return sentence;
}

// This function prints the Grade roundend to the nearest integer.
void print_grade(int index)
{
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 1 && index < 16)
    {
        printf("Grade %i\n", index);
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
}

File: lecture_6/assignements/sentimental-readability/readability.py
from cs50 import get_string


def main():
    # Get the input text
    text = get_string("Text: ")

    # Count each variable
    total_letters = letter_counter(text)
    total_words = word_counter(text)
    total_periods = period_counter(text)

    # Calculate each parameter
    L = total_letters * 100 / total_words
    S = total_periods * 100 / total_words

    # Calculate the readability index
    readability_score = compute_readability_score(L, S)

    # Print readability index
    print_readability_score(readability_score)


# Count the letters in the text
def letter_counter(text):
    counter = 0
    for i in text:
        # Each alphabetic character is a new letter
        if i.isalpha():
            counter += 1
    return counter


# Count the words in the text
def word_counter(text):
    counter = 0
    for i in text:
        # Each space is a new word
        if i.isspace():
            counter += 1
    return counter + 1 # Add 1 to count the last word


# Count the sentences in the text
def period_counter(text):
    counter = 0
    for i in text:
        # Each of this signs means a new sentence
        if i in [".", "!", "?"]:
            counter += 1
    return counter


# Calculate the readability score of the text
def compute_readability_score(L, S):
    # Coleman-Liau formula
    index = 0.0588 * L - 0.296 * S - 15.8
    return index


# Rounds to the nearest integer and print the readability score of the text
def print_readability_score(not_rounded_index):
    index = round(not_rounded_index)

    if index < 1:
        print("Before Grade 1")
    elif 1 <= index < 16:
        print(f"Grade {index}")
    elif index >= 16:
        print("Grade 16+")


main()
